README.md
# Trompa Multimodal Component

> The Multimodal Component React library for the TROMPA project

[![NPM](https://img.shields.io/npm/v/trompa-multimodal-component.svg)](https://www.npmjs.com/package/trompa-multimodal-component) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

This project is part of the [TROMPA project](https://trompamusic.eu).

## Installation

Assuming you already have a React application, the easiest way to start using the Multimodal Component, is to install it
via NPM or Yarn.

If you're using NPM:

```bash
npm install --save trompa-multimodal-component
```

If you're using Yarn:

```bash
yarn add trompa-multimodal-component
```

### Material-UI

The MultiModal Component uses the peer dependency MaterialUI, so make sure you have it installed as well.

NPM:

```bash
npm install @material-ui/core @material-ui/icons
```

Yarn:

```bash
yarn add @material-ui/core @material-ui/icons
```

## Usage

This is a basic usage example. There will be more examples after more options have been added to the Multimodal
Component.

```jsx
import React, { Component } from 'react'
import MultiModalComponent, { SearchConfig, searchTypes } from 'trompa-multimodal-component'

const searchConfig = new SearchConfig({
  searchTypes: [searchTypes.MusicComposition],
});

class Example extends Component {
  render() {
    return (
      <MultiModalComponent
        config={searchConfig}
        onResultClick={node => console.log('User has clicked on:', node)}
        i18n={{
          'en-US': { searchBar: { placeholder_text: 'Search for all music compositions...' } },
          'nl-NL': { searchBar: { placeholder_text: 'Zoek voor alle muziek composities...' } },
        }}
      />
    )
  }
}
```

### Explicit filtering example

```jsx
const searchConfig = new SearchConfig({
  searchTypes: [searchTypes.DigitalDocument],
  fixedFilter: {
    workExample_some: {
      identifier_in: ['56667d40-aa92-4106-a97e-9b28656c56e3', '7e36728d-6112-4deb-9240-1ff77e219f96', '25cdbfcb-af80-4fe6-9dc5-1ae47bcea5e0', 'b31c303e-3484-4787-896c-3c1048995103'],
    },
  },
});
```

### Props

| Prop | Type | Default value | Description | Required |
|------|------|---------------|-------------|----------|
| **searchConfig** | SearchConfig | undefined  | An instance of the SearchConfig class | Yes |
| **onResultClick** | Function | function(**result**: *Object*) { }  | Callback when the user clicks on a result. | No |
| **renderSearchResult** | Function | function(**type**: *String*, **item**: *Object*, **onResultClick**: *function*) { } | Custom render function for search results. | No |
| **placeholderText (deprecated, use i18n)** | String | Enter a search phrase... | Placeholder text for the search input | No |
| **i18n** | Object | undefined | Override translations | No |

#### Currently supported searchTypes

```jsx
const searchTypes = ['AudioObject', 'DigitalDocument', 'Person', 'MusicComposition', 'VideoObject'];
```

#### Custom searchTypes

A custom search type can also be used. This can be either a Class or Object with at least the following properties:
name, filters, searchAllQuery and searchQuery.

For example to create a custom search type for SoftwareApplications;

```js
import gql from 'graphql-tag';

class CustomType {
  static name = 'SoftwareApplication';

  static filters = [];

  static searchAllQuery = gql`
    query($query: String!, $first: Int = 9999) {
      allResults: searchMetadataText(onTypes: [SoftwareApplication], onFields: [title], substring: $query, first: $first) {
        ... on SoftwareApplication {
          identifier
          format
          _searchScore
        }
      }
    }
  `;

  static searchQuery = gql`
    query($filter: _SoftwareApplicationFilter) {
      results: SoftwareApplication(filter: $filter, first: 50) {
        __typename
        ... on SoftwareApplication {
          identifier
          name
          title
          creator
          source
        }
      }
    }
  `;
}
```

If your setup doesn't
support [static class properties](https://babeljs.io/docs/en/babel-plugin-proposal-class-properties), you can define the
properties like this as well:

```js
class CustomType {
}

// or const CustomType = {}

CustomType.name = 'SoftwareApplication';
CustomType.filters = [];
CustomType.searchAllQuery = gql`
    query($query: String!, $first: Int = 9999) {
      allResults: searchMetadataText(onTypes: [SoftwareApplication], onFields: [title], substring: $query, first: $first) {
        ... on SoftwareApplication {
          identifier
          format
          _searchScore
        }
      }
    }
  `;

CustomType.searchQuery = gql`
    query($filter: _SoftwareApplicationFilter) {
      results: SoftwareApplication(filter: $filter, first: 50) {
        __typename
        ... on SoftwareApplication {
          identifier
          name
          title
          creator
          source
        }
      }
    }
  `;
```

#### Custom result

If you want to change the rendered result, you can use the `renderSearchResult` prop to use a custom function. This function should return a valid JSX object.

You can use the `SearchResult` component to quickly customise or add a custom search result. However, you can return any
custom styled component. Make sure to pass the onClick function so that the trompa-multimodal-component can handle this
event accordingly.

```jsx
import React, { Component } from 'react'
import MultiModalComponent, { SearchConfig, searchTypes, SearchResult } from 'trompa-multimodal-component'

const searchConfig = new SearchConfig({
  searchTypes: [searchTypes.MusicComposition],
});

const renderSearchResult = (type, item, onClick) => {
  // You can return custom JSX
  // return (
  //   <div onClick={() => onClick(item)}>{item.title}</div>
  // );

  // If you conditionally return a search result, make sure to always fallback to a default search result like the
  // following line.
  return <SearchResult title={item.title} variant="default" onClick={() => onClick(item)} />
};

class Example extends Component {
  render() {
    return (
      <MultiModalComponent
        config={searchConfig}
        onResultClick={node => console.log('User has clicked on:', node)}
        renderSearchResult={renderSearchResult}
        i18n={{
          'en-US': { searchBar: { placeholder_text: 'Search for all music compositions...' } },
          'nl-NL': { searchBar: { placeholder_text: 'Zoek voor alle muziek composities...' } },
        }}
      />
    )
  }
}
```

## Local development

To start the example demo, run

```bash
yarn start:example
```

And visit http://localhost:5050 in a browser

## License

Apache-2.0 ©


CHANGELOG.md
# [1.4.0](https://github.com/Videodock/trompa-multimodal-component/compare/v1.3.0...v1.4.0) (2021-07-23)


### Bug Fixes

* upgrade graphql-tag from 2.11.0 to 2.12.1 ([01146be](https://github.com/Videodock/trompa-multimodal-component/commit/01146beb0e7e5599a955bb8c32ec6f5580fd49d9))


### Features

* **project:** add check licenses github workflow ([224a23d](https://github.com/Videodock/trompa-multimodal-component/commit/224a23dfff537dd4d1e0491cccbf1d3b1ae09dba))
* **project:** add js-green-licenses dependency ([0311512](https://github.com/Videodock/trompa-multimodal-component/commit/0311512384f618dd46be7ec28bf85afd919ac9c0))
* **project:** add js-green-licenses.json ([a38b652](https://github.com/Videodock/trompa-multimodal-component/commit/a38b652d5321ab2020f2cead84ecf321a3b8be3d))
* **project:** postprocess results ([7b64b3c](https://github.com/Videodock/trompa-multimodal-component/commit/7b64b3cf67785783bdbf1ef6024bf364cba87891))



# [1.3.0](https://github.com/Videodock/trompa-multimodal-component/compare/v1.2.1...v1.3.0) (2021-03-12)


### Features

* **project:** add custom search result render prop ([85b1a2b](https://github.com/Videodock/trompa-multimodal-component/commit/85b1a2b5f03bccbe866224d5148101eb5d6784eb))



## [1.2.1](https://github.com/Videodock/trompa-multimodal-component/compare/v1.2.0...v1.2.1) (2021-03-11)


### Bug Fixes

* **project:** show 50 best matching results ([d46f833](https://github.com/Videodock/trompa-multimodal-component/commit/d46f833bf1ba155b6186ffe59935f57881a15118))



# [1.2.0](https://github.com/Videodock/trompa-multimodal-component/compare/v1.1.0...v1.2.0) (2021-03-11)


### Bug Fixes

* **project:** align mui with styles import ([402558d](https://github.com/Videodock/trompa-multimodal-component/commit/402558db463958d38a1f88ff27c8a61d6861f7d1))
* **project:** fix console warning ([77c5a1f](https://github.com/Videodock/trompa-multimodal-component/commit/77c5a1f93f286ddd171435843823dee0eed18274))
* **project:** remove filters from modals ([7cd00aa](https://github.com/Videodock/trompa-multimodal-component/commit/7cd00aaed50f9043529b3ab300e2688a2cb338e1))
* **search:** edit search results container ([293c278](https://github.com/Videodock/trompa-multimodal-component/commit/293c27888f7f3459db23842f4d0ac514219fdfca))


### Features

* **search:** add search results component ([0b246dd](https://github.com/Videodock/trompa-multimodal-component/commit/0b246dd488925af31ec49a886938bd047bbe9a96))
* **search:** add search tips component ([4baaab9](https://github.com/Videodock/trompa-multimodal-component/commit/4baaab9e10a0d6215392c86356d72a855736e5a6))



# [1.1.0](https://github.com/Videodock/trompa-multimodal-component/compare/v1.0.1...v1.1.0) (2021-03-09)


### Features

* **project:** update translations and make them customizable ([64ad73a](https://github.com/Videodock/trompa-multimodal-component/commit/64ad73aedbe6ecfa0933ecb0a9a14fb67a598d1c))



## [1.0.1](https://github.com/Videodock/trompa-multimodal-component/compare/v1.0.0...v1.0.1) (2021-03-02)


### Bug Fixes

* **project:** remove engines from package.json ([c410f85](https://github.com/Videodock/trompa-multimodal-component/commit/c410f8593d542fcc4098da51f2d8f888fc7061c8))



# [1.0.0](https://github.com/Videodock/trompa-multimodal-component/compare/v1.0.0-alpha4...v1.0.0) (2021-02-26)



# [1.0.0-alpha4](https://github.com/Videodock/trompa-multimodal-component/compare/v1.0.0-alpha3...v1.0.0-alpha4) (2020-11-12)


### Bug Fixes

* **search:** fix title and source overlapping parent ([b745c7f](https://github.com/Videodock/trompa-multimodal-component/commit/b745c7fb01a3fd6c13edcc1a29ba69067322cffa))
* **search:** keep filters drawer mounted to prevent state loss ([faa34f9](https://github.com/Videodock/trompa-multimodal-component/commit/faa34f9552106b12d69f85dc66a82b71a563c5fe))



# [1.0.0-alpha3](https://github.com/Videodock/trompa-multimodal-component/compare/v1.0.0-alpha2...v1.0.0-alpha3) (2020-11-09)


### Bug Fixes

* **project:** add name property to search types ([49b22be](https://github.com/Videodock/trompa-multimodal-component/commit/49b22befe3054a3e4a9b36d26b9801d60e46be49))
* **project:** fix filter disappearing when list is empty ([849f483](https://github.com/Videodock/trompa-multimodal-component/commit/849f483f0b3f815d1f232f723b0c484ddad6f273))


### Features

* **project:** add birthplace filter to person search type ([8bd0c9f](https://github.com/Videodock/trompa-multimodal-component/commit/8bd0c9f4343e4d45cda45a97f08f30a0af15e4a4))
* **project:** allow custom placeholder text ([010d113](https://github.com/Videodock/trompa-multimodal-component/commit/010d1131a9e9ae9e6729240b03ba2aa6ea670beb))



# [1.0.0-alpha2](https://github.com/Videodock/trompa-multimodal-component/compare/v1.0.0-alpha1...v1.0.0-alpha2) (2020-11-05)


### Bug Fixes

* **project:** add material ui as peer dependency ([8c22422](https://github.com/Videodock/trompa-multimodal-component/commit/8c22422d03b9a4761822f2046da75e14662926c7))



# [1.0.0-alpha1](https://github.com/Videodock/trompa-multimodal-component/compare/v0.4.0...v1.0.0-alpha1) (2020-11-04)


### Features

* **project:** add new search mechanism ([edc4583](https://github.com/Videodock/trompa-multimodal-component/commit/edc45833e2ab97f2baae0f62736120e1013238f5))
* **project:** update search mechanism to to support use-cases ([2c9cb01](https://github.com/Videodock/trompa-multimodal-component/commit/2c9cb01e05d18992ed8bf6f849198e11358ae21e))


### BREAKING CHANGES

* **project:** this commit changes the MultiModalComponent props.



# [0.4.0](https://github.com/Videodock/trompa-multimodal-component/compare/v0.3.1...v0.4.0) (2020-03-26)


### Bug Fixes

* **project:** add netlify.toml to fix Netlify builds ([1a0782c](https://github.com/Videodock/trompa-multimodal-component/commit/1a0782ca735eefd62b66a0632eaaec37f58b13b6))

example/App.js
import React, { useState } from 'react';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import Dialog from '@material-ui/core/Dialog';
import Paper from '@material-ui/core/Paper';
import Switch from '@material-ui/core/Switch';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import gql from 'graphql-tag';
import MultiModalComponent, { SearchConfig, searchTypes, SearchResult } from '../src/index';

class CustomType {
  static name = 'SoftwareApplication';

  static filters = [];

  static searchAllQuery = gql`
    query($query: String!, $first: Int = 9999) {
      allResults: searchMetadataText(onTypes: [SoftwareApplication], onFields: [title], substring: $query, first: $first) {
        ... on SoftwareApplication {
          identifier
          format
          _searchScore
        }
      }
    }
  `;

  static searchQuery = gql`
    query($filter: _SoftwareApplicationFilter) {
      results: SoftwareApplication(filter: $filter, first: 50) {
        __typename
        ... on SoftwareApplication {
          identifier
          name
          title
          creator
          source
        }
      }
    }
  `;
}

class PostprocessCustomType {
  static name = 'MusicComposition';

  static filters = [];

  static preprocessQuery = query => {
    return `(?i).*${query}.*`;
  }

  static searchAllQuery = gql`
    query($query: String!) {
      allResults: ItemList(identifier:"e91489d7-a776-40dd-8abf-0c934922bd99") {
        identifier
        name
        itemListElement(filter:{name_regexp:$query}) {
          identifier
          name
        }
      }
    }
  `;

  static searchQuery = gql`
    query($filter: _ThingInterfaceFilter) {
      results: ItemList(identifier:"e91489d7-a776-40dd-8abf-0c934922bd99") {
        identifier
        name
        itemListElement(filter: $filter, first: 50) {
          identifier
          name
        }
      }
    }
  `;

  static processSearchResult = result => {
    // Find the itemListElements of this ItemList, instead of a list of ItemLists
    if (Array.isArray(result) && result[0]) {
      return result[0].itemListElement;
    }

    return [];
  }
}

const BlockQuote = ({ children }) => {
  return (
    <div style={{ backgroundColor: '#fff', borderLeft: `6px solid rgb(63 81 181)`, padding: 16, marginBottom: 8 }}>
      <Typography variant="caption">{children}</Typography>
    </div>
  );
};

const MultiModalComponentSelect = ({ config, i18n, placeholderText, production, renderSearchResult }) => {
  const [open, setOpen]         = useState(false);
  const [selected, setSelected] = useState();

  return (
    <React.Fragment>
      <div className="mmc-select">
        <Typography className="mmc-select-value">{selected ? selected.name : 'No selection'}</Typography>
        <Button variant="contained" size="small" color="primary" onClick={() => setOpen(true)}>
          Select
        </Button>
      </div>
      <Dialog PaperProps={{ style: { width: '100%' } }} open={open} keepMounted={false} onClose={() => setOpen(false)} maxWidth="md">
        <MultiModalComponent
          uri={production ? 'https://api.trompamusic.eu' : 'https://api-test.trompamusic.eu'}
          config={config}
          placeholderText={placeholderText}
          renderSearchResult={renderSearchResult}
          i18n={i18n}
          onResultClick={item => {
            setSelected(item);
            setOpen(false);
          }}
        />
      </Dialog>
    </React.Fragment>
  );
};

const uc1Config = new SearchConfig({
  searchTypes: [searchTypes.DigitalDocument],
  fixedFilter: {
    workExample_some: {
      identifier_in: ['56667d40-aa92-4106-a97e-9b28656c56e3', '7e36728d-6112-4deb-9240-1ff77e219f96', '25cdbfcb-af80-4fe6-9dc5-1ae47bcea5e0', 'b31c303e-3484-4787-896c-3c1048995103'],
    },
  },
});

const uc2Config = new SearchConfig({
  searchTypes: [searchTypes.AudioObject],
  fixedFilter: {
    creator_starts_with: 'https://www.voiceful.io',
    exampleOfWork      : { identifier: 'f7bd0cfc-0c34-4204-8435-47d81a825640' },
  },
});

const uc3Config = new SearchConfig({
  searchTypes: [searchTypes.DigitalDocument],
  fixedFilter: {
    workExample_some: { identifier: '56667d40-aa92-4106-a97e-9b28656c56e3' },
  },
});

const ex1Config = new SearchConfig({
  searchTypes: [searchTypes.Person],
});

const ex2Config = new SearchConfig({
  searchTypes: [searchTypes.AudioObject, searchTypes.VideoObject],
});

const ex3Config = new SearchConfig({
  searchTypes: [searchTypes.MusicComposition],
});

const ex4Config = new SearchConfig({
  searchTypes: [searchTypes.DigitalDocument],
});

const ex5Config = new SearchConfig({
  searchTypes: [CustomType],
});

const ex6Config = new SearchConfig({
  searchTypes: [CustomType, searchTypes.MusicComposition],
});

const ex7Config = new SearchConfig({
  searchTypes: [PostprocessCustomType],
});

const App = () => {
  const [production, setProduction] = useState(false);

  return (
    <section className="section">
      <FormControlLabel
        control={
          <Switch
            checked={production}
            onChange={() => setProduction(!production)}
            name="production"
            color="primary"
          />
        }
        label="Production"
      />
      <Paper style={{ padding: 16, backgroundColor: '#f1f1f1', marginBottom: 64 }} color="red" variant="outlined">
        <Typography variant="h6" gutterBottom>Examples:</Typography>
        <BlockQuote>
          As a user I want to be able to find a single type (Person) with related facets.
        </BlockQuote>
        <MultiModalComponentSelect
          config={ex1Config}
          i18n={{
            'en-US': { searchBar: { placeholder_text: 'Search for Persons in the CE' } },
            'nl-NL': { searchBar: { placeholder_text: 'Zoek naar Personen in de CE' } },
          }}
          production={production}
        />
        <BlockQuote>
          As a user I want to be able to find multiple types (AudioObject and VideoObject) with related facets.
        </BlockQuote>
        <MultiModalComponentSelect config={ex2Config} placeholderText="Search for Music and Video recordings in the CE" production={production} />
        <BlockQuote>
          As a user I want to be able to find music compositions with related facets.
        </BlockQuote>
        <MultiModalComponentSelect config={ex3Config} production={production} />
        <BlockQuote>
          As a user I want to be able to find scores with related facets.
        </BlockQuote>
        <MultiModalComponentSelect config={ex4Config} placeholderText="Search for scores" production={production} />
        <BlockQuote>
          As a user I want to be able to find software applications using a custom type.
        </BlockQuote>
        <MultiModalComponentSelect
          config={ex5Config}
          placeholderText="Search for software applications"
          production={production}
          renderSearchResult={(type, item, onClick) => <SearchResult type="SOFTWARE" title={item.title} variant="default" onClick={() => onClick(item)} />}
        />
        <BlockQuote>
          As a developer I want to be able to customise the search results
        </BlockQuote>
        <MultiModalComponentSelect
          config={ex6Config}
          placeholderText="Search for software applications and MusicCompositions"
          production={production}
          renderSearchResult={(type, item, onClick) => {
            if (type === CustomType.name) {
              return <div onClick={() => onClick(item)}>CustomType: {item.title}</div>;
            }

            return <div onClick={() => onClick(item)}>MusicComposition: {item.title}</div>;
          }}
        />
        <BlockQuote>
          As a developer I want to be able to post-process search results.
        </BlockQuote>
        <MultiModalComponentSelect config={ex7Config} production={production} />
      </Paper>
      <Paper style={{ padding: 16, backgroundColor: '#f1f1f1', marginBottom: 64 }} color="red" variant="outlined">
        <Typography variant="h6" gutterBottom>Use cases:</Typography>
        <BlockQuote>
          Give me all scores with synth audios: I can find musical scores (DigitalDocument) with associated synthesized audios (AudioObject with creator: “https://www.voiceful.io/”).
        </BlockQuote>
        <MultiModalComponentSelect config={uc1Config} placeholderText="Search for scores with associated synthesized audios" production={production} />
        <BlockQuote>
          I can find synthesized audios (AudioObject with creator: “https://www.voiceful.io/”) for a given score, DigitalDocument with id: “x”
        </BlockQuote>
        <MultiModalComponentSelect config={uc2Config} placeholderText="Search for audio recordings for a given score" production={production} />
        <BlockQuote>
          I have this synth audio, give me the score from which it was created: I can find musical score (DigitalDocument) from which a synthesized audio, AudioObject with creator: “https://www.voiceful.io/”, was created.
        </BlockQuote>
        <MultiModalComponentSelect config={uc3Config} placeholderText="Search for a score" production={production} />
      </Paper>
    </section>
  );
};

export default App;
